Pal/security/linux/graphene-ioctl.c

```diff
@@ -0,0 +1,380 @@
+	stattmp.st_atime	= stat.atime.tv_sec;
+	stattmp.st_mtime	= stat.mtime.tv_sec;
+	stattmp.st_ctime	= stat.ctime.tv_sec;
+
+	return copy_to_user(statbuf, &stattmp, sizeof(stat)) ? -EFAULT : 0;
+}
+
+long grm_sys_unlink (struct graphene_info *gi, const char __user *filename)
+{
+	return -ENOSYS;
+}
+
+long grm_sys_rmdir (struct graphene_info *gi, const char __user *filename)
+{
+	return -ENOSYS;
+}
+
+long grm_sys_bind (struct graphene_info *gi,
+		   int sockfd, struct sockaddr __user *addr, int addrlen)
+{
+	struct socket *sock;
+	struct sockaddr_storage address;
+	int error;
+
+	if (addrlen < 0 || addrlen > sizeof(struct sockaddr_storage))
+		return -EINVAL;
+	if (addrlen == 0)
+		return 0;
+	if (copy_from_user(&address, addr, addrlen))
+		return -EFAULT;
+
+	sock = sockfd_lookup(sockfd, &error);
+	if (!sock)
+		return error;
+
+	/* Reference monitor: check "address" */
+	if (gi) {
+		error = check_bind_addr(gi, sock, (struct sockaddr *) &address,
+					addrlen);
+		if (error < 0) {
+			sockfd_put(sock);
+			return error;
+		}
+	}
+
+	error = kernel_bind(sock, (struct sockaddr *) &address, addrlen);
+	sockfd_put(sock);
+	return error;
+}
+
+long grm_sys_connect (struct graphene_info *gi,
+		      int sockfd, struct sockaddr __user *addr, int addrlen)
+{
+	struct socket *sock;
+	struct sockaddr_storage address;
+	int error;
+
+	if (addrlen < 0 || addrlen > sizeof(struct sockaddr_storage))
+		return -EINVAL;
+	if (addrlen == 0)
+		return 0;
+	if (copy_from_user(&address, addr, addrlen))
+		return -EFAULT;
+
+	sock = sockfd_lookup(sockfd, &error);
+	if (!sock)
+		return error;
+
+	/* Reference monitor: check "address" */
+	if (gi) {
+		error = check_connect_addr(gi, sock,
+					   (struct sockaddr *) &address,
+					   addrlen);
+		if (error < 0) {
+			sockfd_put(sock);
+			return error;
+		}
+	}
+
+	error = kernel_connect(sock, (struct sockaddr *) &address, addrlen,
+			       sock->file->f_flags);
+	sockfd_put(sock);
+	return error;
+}
+
+long grm_sys_execve (struct graphene_info *gi,
+		     const char __user *const __user *argv)
+{
+	struct file *exe_file = get_task_exe_file(current);
+	struct filename *tmp, *new;
+	int len = 0;
+	const char *path;
+
+	tmp = __getname();
+	if (!tmp)
+		return -ENOMEM;
+
+	path = dentry_path_raw(exe_file->f_path.dentry,
+			       (char *) tmp, PATH_MAX);
+	if (IS_ERR(path)) {
+		__putname(tmp);
+		return PTR_ERR(path);
+	}
+
+	len = strlen(path);
+	new = kmalloc(sizeof(*new), GFP_KERNEL);
+	if (!new) {
+		__putname(tmp);
+		return -ENOMEM;
+	}
+
+	new->name = __getname();
+	if (!new->name) {
+		__putname(tmp);
+		kfree(new);
+		return -ENOMEM;
+	}
+
+	strncpy((char *) new->name, path, len + 1);
+	init_filename(new);
+	__putname(tmp);
+
+#ifdef GRAPHENE_DEBUG
+	printk(KERN_INFO "Graphene: PID %d EXEC %s\n",
+	       current->pid, new->name);
+#endif
+
+	return KSYM(do_execve)(new, argv, NULL);
+}
+
+static long grm_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+{
+	char data[256];
```

Comment:
Should be `_IOC_SIZEMASK` instead of `256`, otherwise this buffer can be overflowed with usermode-controlled data (see line 259, with `copy_from_user` call).